### DISCUSSION TITLE
Creating API from cookiecutter template.

### DISCUSSION THREAD

**Discussion by DevilsAutumn** [cookiecutter/cookiecutter-django#4253] [Q&A] [answered]
Hi all, I want to create an api using django , I just wanted to know what modification do i need to do in cookiecutter template to do so and host the api on aws?
Also Do i need traefik even for apis?
Thankyou!

**Comment by foarsitter** [comment, marked as answer]
The Python language itself is an [API](https://docs.python.org/3/c-api/index.html). Since you mention AWS I assume you want to create a HTTP API? When serving a HTTP API you need some sort of loadbalancer for your app. In this template Traefik routes you http traffic to gunicorn which serves your wsig application. I'm not enough familiar with the AWS ecosystem but I think they have an service for that (Elastic Beanstalk maybe?).

This template comes with an HTTP REST API included when using `use_drf=yes`, it uses https://www.django-rest-framework.org/ for that. The user app has some API endpoints: https://github.com/cookiecutter/cookiecutter-django/blob/1cba3b62e04a79d61cde9e90b2987ba3630a03c6/%7B%7Bcookiecutter.project_slug%7D%7D/%7B%7Bcookiecutter.project_slug%7D%7D/users/api/views.py#L13-L20

My advice is to start with one thing when learning new things. So if you are new to Python spend some time learning Python. When you are new to Django learn Django. When you have some experience with Python & Django start with the rest framework. Each tool adds a layer of complexity. The cookiecutter template adds a LOT of layers. Each dependency comes with some sort of a configuration and each configuration can go wrong (I learned the hard way spending hours of debugging) so take it easy in the beginning.

**Reply by DevilsAutumn** [reply]
Thanks for the reply! I am familiar with python, django and drf but cookiecutter has a lot many complex layers which will take time. Probably I should start with a basic setup and then add things one by one.

**Reply by foarsitter** [reply]
Browsed through your github profile and it indeed seems you are familiar with Django :) Dive in I would say, give it a try!

**Reply by DevilsAutumn** [reply]
hehe :) Actually the main problem i'm facing is setting up traefik . The error is same as mentioned in [this discussion](https://github.com/cookiecutter/cookiecutter-django/discussions/3555). What i want to do is that my ec2 's public ipv4 dns (for ex:- ec2-26-76-98-168.eu-north-2.compute.amazonaws.com) should work on https so that i can make api calls from my frontend which is hosted somewhere else with my domain pointing to the frontend. But setting up traefik requries some configuration in the domain's dns records if i'm not wrong.
Does it sound reasonable to you? 
Sorry if i sound wierd, I haven't deployed any app at this scale (and on aws) yet😅 but with some help , i can definitely do it!

**Reply by foarsitter** [reply]
Do you have control over a different domain? Maybe you can try that since I have no idea how amazon handles routing. 

Try resolve the domain https://toolbox.googleapps.com/apps/dig/#A/ (both A  & AAAA). Does traefik respond when accessed when you try the ip address of the server?

**Reply by DevilsAutumn** [reply]
>Do you have control over a different domain?

yes i bought one from godaddy.

I read on [SO](https://stackoverflow.com/questions/66686686/enable-https-on-ec2-public-dns) that we cannot add ssl to public dns of ec2 instance instead we have to use a third party service and a domain. So setting up traefik seems to be a way forward.

Do you have any idea what settings are required for setting up traefik ?

**Comment by DevilsAutumn** [comment]
Hey Just wanted to let you know that i successfully configured traefik and everything is working fine now :)
Thanks for the help!

**Reply by foarsitter** [reply]
Nice one! To which length did you need to go to solve it? Are there any thing helpful for future reference?

**Reply by DevilsAutumn** [reply]
To setup the project with docker, traefik , nginx, postgres on aws ec2, I had to do the following things:
1. First i  created an ec2 instance on aws .
2. Then i created IAM role and assigned it to the ec2 instance.(Credentials can also be used)
3. Then I created an security group and added inbound rule to allow https traffic in it.
4. Then i created an elastic IP and associated it with my instance so that my instance has a fixed public IP, even if the instance reboots.
5. To Setup traefik i copied the elastic IP and pasted it in A record of my DNS settings on godaddy.
Everything started working fine after that :)
